tests/utils/managed_utils/mod.rs

```diff
@@ -0,0 +1,583 @@
+    pub group_id: GroupId,
+    pub members: Vec<(usize, Vec<u8>)>,
+    pub ciphersuite: Ciphersuite,
+    pub group_config: ManagedGroupConfig,
+    pub public_tree: Vec<Option<Node>>,
+}
+
+impl Group {
+    pub fn random_group_member(&self) -> Vec<u8> {
+        let index = (OsRng.next_u32() as usize) % self.members.len();
+        // We can unwrap here, because the index is scoped with the size of the
+        // HashSet.
+        let (_, identity) = self.members.iter().nth(index).unwrap().clone();
+        identity
+    }
+}
+
+pub enum ActionType {
+    Commit,
+    Proposal,
+}
+
+pub enum OperationType {
+    Update,
+    Add,
+    Remove,
+}
+
+pub struct ManagedTestSetup<'client_lifetime> {
+    pub number_of_clients: usize,
+    // The clients identity is its position in the vector in be_bytes.
+    pub clients: RefCell<HashMap<Vec<u8>, RefCell<Client<'client_lifetime>>>>,
+    pub groups: RefCell<HashMap<GroupId, Group>>,
+    pub key_store: KeyStore,
+    // This maps key package hashes to client ids.
+    pub waiting_for_welcome: RefCell<HashMap<Vec<u8>, Vec<u8>>>,
+    pub default_mgc: ManagedGroupConfig,
+}
+
+// Some notes regarding the layout of the `ManagedTestSetup` implementation
+// below: The references to the credentials in the ManagedGroups (in the
+// clients) are essentially references to the keystore, which in turns means
+// they are references to the top-level object, i.e. the setup.
+//
+// As a result, we can't have mutable references to the setup, because we have
+// living (immutable) references to the keystore floating around. As a follow-up
+// result, we need to populate the keystore _completely_ before distributing the
+// references, as we can't mutably reference the keystore.
+//
+//   * Note, that (to my knowledge) we can't have the keystore live in a
+//   refcell, because then the references are based on the local `borrow()`ed
+//   object and won't live long enough.
+//
+// Finally, this means we have to initialize the KeyStore before we create the
+// ManagedTestSetup object and we can create the clients (which contain the
+// references to the KeyStore) only after we do that. This has to happen in the
+// context that the `ManagedTestSetup` lives in, because otherwise the
+// references don't live long enough.
+
+impl<'ks> ManagedTestSetup<'ks> {
+    /// Create a new `ManagedTestSetup` with the given default
+    /// `ManagedGroupConfig` and the given number of clients. For lifetime
+    /// reasons, `create_clients` has to be called in addition with the same
+    /// number of clients.
+    pub fn new(default_mgc: ManagedGroupConfig, number_of_clients: usize) -> Self {
+        let mut key_store = KeyStore {
+            credential_bundles: HashMap::new(),
+        };
+        // Create credentials first to avoid borrowing issues.
+        for i in 0..number_of_clients {
+            let identity = i.to_be_bytes().to_vec();
+            // For now, everyone supports all ciphersuites.
```

Comment:
Another thing to extend later.

Reply:
Yes, I'll add an issue for that once this PR lands.